synapseclient/core/download/download_functions.py

```diff
@@ -0,0 +1,727 @@
+    SynapseMd5MismatchError,
+    SynapseError,
+)
+from synapseclient.core.remote_file_storage_wrappers import S3ClientWrapper, SFTPWrapper
+from synapseclient.core.retry import (
+    DEFAULT_RETRY_STATUS_CODES,
+    RETRYABLE_CONNECTION_ERRORS,
+    RETRYABLE_CONNECTION_EXCEPTIONS,
+    with_retry,
+)
+
+from synapseclient.core.utils import (
+    MB,
+)
+
+if TYPE_CHECKING:
+    from synapseclient import Synapse, Entity
+
+FILE_BUFFER_SIZE = 2 * MB
+REDIRECT_LIMIT = 5
+
+# Defines the standard retry policy applied to the rest methods
+# The retry period needs to span a minute because sending messages is limited to 10 per 60 seconds.
+STANDARD_RETRY_PARAMS = {
+    "retry_status_codes": DEFAULT_RETRY_STATUS_CODES,
+    "retry_errors": RETRYABLE_CONNECTION_ERRORS,
+    "retry_exceptions": RETRYABLE_CONNECTION_EXCEPTIONS,
+    "retries": 60,  # Retries for up to about 30 minutes
+    "wait": 1,
+    "max_wait": 30,
+    "back_off": 2,
+}
+
+
+async def download_file_entity(
+    *,
+    download_location: str,
+    entity: "Entity",
+    if_collision: str,
+    submission: str,
+    synapse_client: Optional["Synapse"] = None,
+) -> None:
+    """
+    Download file entity
+
+    Arguments:
+        download_location: The download location
+        entity:           The Synapse Entity object
+        if_collision:      Determines how to handle file collisions.
+                            May be
+
+            - `overwrite.local`
+            - `keep.local`
+            - `keep.both`
+
+        submission:       Access associated files through a submission rather than through an entity.
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    # set the initial local state
+    entity.path = None
+    entity.files = []
+    entity.cacheDir = None
+
+    # check to see if an UNMODIFIED version of the file (since it was last downloaded) already exists
+    # this location could be either in .synapseCache or a user specified location to which the user previously
+    # downloaded the file
+    cached_file_path = client.cache.get(
+        file_handle_id=entity.dataFileHandleId, path=download_location
+    )
+
+    # location in .synapseCache where the file would be corresponding to its FileHandleId
+    synapse_cache_location = client.cache.get_cache_dir(
+        file_handle_id=entity.dataFileHandleId
+    )
+
+    file_name = (
+        entity._file_handle.fileName
+        if cached_file_path is None
+        else os.path.basename(cached_file_path)
+    )
+
+    # Decide the best download location for the file
+    if download_location is not None:
+        # Make sure the specified download location is a fully resolved directory
+        download_location = ensure_download_location_is_directory(download_location)
+    elif cached_file_path is not None:
+        # file already cached so use that as the download location
+        download_location = os.path.dirname(cached_file_path)
```

Comment:
Would this lead to overwriting the file without warning?

Reply:
> Would this lead to overwriting the file without warning?

Not in this case - This first check is "If I were to download the file, where _would_ I download that file to". Later on the `resolve_download_path_collisions` handles for this.

> Also I remember you mentioning something about os.path not working for Windows because of the \ delimiter, is this a concern in this context?

It's not a problem in this case because the script running on a windows machine is going to have drives mapped with the correct delimiter with `os.path`.